### PR TITLE
Fix decimal stock quantities below 1 being marked out of stock

### LINKED ISSUE #41676: Stock Quantity cannot be set below zero using filters because of integer casting of get_stock_quantity()
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Decimal values below zero like 0.4 or 0.5 work only till WC 7.7.2.

From WooCommerce 7.8+, we see an "Out of Stock" error when the stock amount is overridden using the below filter:

```
// Removes the WooCommerce filter, that is validating the quantity to be an int
remove_filter('woocommerce_stock_amount', 'intval');
  
// Add a filter, that validates the quantity to be a float
add_filter('woocommerce_stock_amount', 'floatval');
```

You will an Out of Stock Error:

image

This PR is causing this: https://github.com/woocommerce/woocommerce/pull/37855/files

Integer casting of `$this->get_stock_quantity()` can be removed to allow merchants/plugins to use decimal values below 0.

More details here: p1700211395088159-slack-C7U3Y3VMY

### Expected behavior

Removing Integer casting resolves this on WC 7.8+

### Actual behavior

When merchants or plugins override the default stock amount, they see "out of stock" message if the quantity goes below 0.

### Steps to reproduce

1. On WC 7.7.2, add the above snippet
2. Set stock quantity to 0.5, See everything works fine 
3. On WC 7.8, add the above snippet
2. Set stock quantity to 0.5, see the out of stock message 

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://magenta-manatee.jurassic.ninja
Site address (URL): https://magenta-manatee.jurassic.ninja
WC Version: 8.3.1
REST API Version: ✔ 8.3.1
WC Blocks Version: ✔ 11.4.9
Action Scheduler Version: ✔ 3.6.4
Log Directory Writable: ✔
WP Version: 6.4.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.58 (Unix) OpenSSL/1.0.2g
PHP Version: 8.0.30
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.3.1
WC Database Prefix: wp_
Total Database Size: 4.28MB
Database Data Size: 2.61MB
Database Index Size: 1.67MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.52MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 1
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (3) ###

Companion Plugin: by Osk – 1.30
Jetpack: by Automattic – 12.8.1
WooCommerce: by Automattic – 8.3.1

### Inactive Plugins (2) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3
Hello Dolly: by Matt Mullenweg – 1.7.2

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /srv/users/user21407a4b/apps/user21407a4b/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/error.php
/srv/users/user21407a4b/apps/user21407a4b/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/notice.php
/srv/users/user21407a4b/apps/user21407a4b/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/success.php

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-virtual-downloadable
product-external-affiliate
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-11-25 13:21:37 +00:00
Options: ✔
Notes: 61
Onboarding: skipped

### Action Scheduler ###

Pending: 2
Oldest: 2023-11-24 13:21:50 +0000
Newest: 2023-11-24 13:22:38 +0000

### Status report information ###

Generated at: 2023-11-24 13:22:15 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by ayushpahwa**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

`WC_Abstract_Product::validate_props()` decides whether a stock-managed product should be flipped to "Out of stock" by comparing `(int) $this->get_stock_quantity()` against `woocommerce_notify_no_stock_amount`. The `(int)` cast truncates any positive fractional stock (e.g. `0.5`) to `0`, which is not above the default threshold (`0`), so a product whose stock the merchant deliberately set to `0.5` via the documented `woocommerce_stock_amount` -> `floatval` filter pair was marked out of stock on save.

This PR replaces the `(int)` cast with a `(float)` cast on both sides of the comparison, so decimal stock quantities are preserved. Integer stock quantities continue to behave identically.

Closes #41676.

(For Bug Fixes) Bug introduced in PR #37855.

### How to test the changes in this Pull Request:

1. In a clean Woo install, add this MU plugin to allow float stock quantities:

   ```php
   <?php
   add_action( 'plugins_loaded', function () {
       remove_filter( 'woocommerce_stock_amount', 'intval' );
       add_filter( 'woocommerce_stock_amount', 'floatval' );
   }, 5 );
   ```

2. Create a Simple product. Enable stock management. Set stock quantity to `0.5`. Save.
3. Visit the product page on the storefront.
4. Before this fix: the product shows "Out of stock" and no Add to Cart button.
5. After this fix: the product is in stock and the Add to Cart button is visible.

### Testing that has already taken place:

- Added a PHPUnit regression test in `WC_Abstract_Product_Test` covering the float-stock-quantity scenario and the zero-stock baseline.
- Ran `pnpm --filter=@woocommerce/plugin-woocommerce lint:php:changes` and `pnpm --filter=@woocommerce/plugin-woocommerce lint:changes:branch` -> clean.
- Ran `composer exec -- phpstan analyse includes/abstracts/abstract-wc-product.php --memory-limit=2G` -> no errors.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

### Changelog entry

Changelog entry created manually at `plugins/woocommerce/changelog/fix-rsmapgj-338`.

<details>
[pipeline] rsmapgj impl rev 1
</details>

### CODE REVIEW

**Comment by ayushpahwa** [comment]
Hey @jorgeatorres, assigned you for review request since you reviewed the related pr earlier. Please feel free to reassign, thanks!

File: plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php [hunk 1]
```diff
@@ -338,4 +338,46 @@ protected function adjust_cogs_value_before_set( ?float $value ): ?float {
...
 		$this->assertEquals( 123.4, $product->get_cogs_value() );
 	}
+
+	/**
+	 * @testdox validate_props() keeps a product in stock when stock quantity is a positive float below 1 and the woocommerce_stock_amount filter is set to floatval.
+	 *
+	 * Regression test for https://github.com/woocommerce/woocommerce/issues/41676 — the
+	 * (int) cast on get_stock_quantity() inside validate_props() truncated 0.5 to 0,
+	 * which is not above the default no-stock threshold (0), so the product was
+	 * incorrectly flipped to "outofstock".
```
Comment (jorgeatorres) [inline plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php]: I think it should be safe to remove this extra info. Maybe just link to the issue "for more details".

File: plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php [hunk 2]
```diff
@@ -338,4 +338,46 @@ protected function adjust_cogs_value_before_set( ?float $value ): ?float {
...
+	 */
+	public function test_validate_props_preserves_in_stock_for_float_stock_quantity() {
+		remove_all_filters( 'woocommerce_stock_amount' );
```
Comment (jorgeatorres) [inline plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php]: Not a problem at the moment, but why remove all filters? I think we could just remove the `intval` one, which is the one we're adding back at the end anyways.

File: plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php [hunk 3]
```diff
@@ -338,4 +338,46 @@ protected function adjust_cogs_value_before_set( ?float $value ): ?float {
...
+		add_filter( 'woocommerce_stock_amount', 'floatval' );
+
+		try {
```
Comment (jorgeatorres) [inline plugins/woocommerce/tests/php/includes/abstracts/class-wc-abstract-product-test.php]: No need to enclose the check in try/finally.

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @ayushpahwa!

Thanks for the PR. Overall this looks good to me. Left some minor feedback, and I'd appreciate if you could also add a test that checks that setting `woocommerce_notify_no_stock_amount` to a non-zero value (i.e. different from the default) also works correctly? We're currently only checking the stock amount while leaving the threshold untouched.

Thanks!

File: plugins/woocommerce/includes/abstracts/abstract-wc-product.php [hunk 4]
```diff
@@ -1527,7 +1527,7 @@ public function validate_props() {
 			return;
 		}
 
-		$stock_is_above_notification_threshold = ( (int) $this->get_stock_quantity() > absint( get_option( 'woocommerce_notify_no_stock_amount', 0 ) ) );
+		$stock_is_above_notification_threshold = ( (float) $this->get_stock_quantity() > (float) get_option( 'woocommerce_notify_no_stock_amount', 0 ) );
```
Comment (jorgeatorres) [inline plugins/woocommerce/includes/abstracts/abstract-wc-product.php]: For consistency with the previous code and a bit more safety, could we pass the float for `woocommerce_notify_no_stock_amount` through `abs()`?

**Review by jorgeatorres — COMMENTED** [review]
@ayushpahwa: thanks for making the changes. I left one more (tiny) suggestion.

Note that, overall, fractional stock is something we don't support very well and there are lots of places in the code where we assume the stock is integer or convert it to integer (such as in `wc_trigger_stock_change_notifications()` or `wc_trigger_stock_change_actions()`, and also some stats and reports). I don't think right now is the moment for a full treatment of this, but thought it was worth flagging. This PR would at least bring the behavior that existed before PR #37855, which is useful.

File: plugins/woocommerce/includes/abstracts/abstract-wc-product.php (hunk 4 above)
Comment (ayushpahwa) [reply plugins/woocommerce/includes/abstracts/abstract-wc-product.php]: Good catch, I've updated the code and added a test!

**Review by jorgeatorres — APPROVED** [review]
